boardswarm/src/hifive_p550_storage.rs

```diff
@@ -0,0 +1,209 @@
+use tokio::{fs, io::AsyncWriteExt, time};
+use tokio_stream::StreamExt;
+use tracing::{debug, instrument, trace, warn};
+
+use crate::{
+    registry::{self, Properties},
+    udev::{DeviceEvent, DeviceRegistrations, PreRegistration},
+    Server, Volume, VolumeError, VolumeTarget, VolumeTargetInfo,
+};
+
+pub const PROVIDER: &str = "hifive-p550-storage";
+
+const MOUNT_TIMEOUT: Duration = Duration::from_secs(10);
+const MOUNT_INTERVAL: Duration = Duration::from_secs(1);
+const TARGETS: [(&str, &str); 1] = [("bootloader", "bootloader_ddr5_secboot.bin")];
+
+#[derive(Deserialize, Debug, Default)]
+struct HifiveP550StorageParameters {
+    #[serde(rename = "match")]
+    match_: HashMap<String, String>,
+}
+
+#[instrument(skip(server, parameters))]
+pub async fn start_provider(name: String, parameters: Option<serde_yaml::Value>, server: Server) {
+    let registrations = DeviceRegistrations::new(server);
+    let provider_properties = &[
+        (registry::PROVIDER_NAME, name.as_str()),
+        (registry::PROVIDER, PROVIDER),
+    ];
+    let mut devices = crate::udev::DeviceStream::new("block").unwrap();
+    let parameters: HifiveP550StorageParameters = if let Some(parameters) = parameters {
+        serde_yaml::from_value(parameters).unwrap()
+    } else {
+        Default::default()
+    };
+    while let Some(d) = devices.next().await {
+        match d {
+            DeviceEvent::Add { device, seqnum } => {
+                let mut properties = device.properties(&name);
+                if !properties.matches(&parameters.match_) {
+                    debug!(
+                        "Ignoring device {} - {:?}",
+                        device.syspath().display(),
+                        properties,
+                    );
+                    continue;
+                }
+                properties.extend(provider_properties);
+
+                let prereg = registrations.pre_register(&device, seqnum);
+                tokio::spawn(async move {
+                    if let Err(e) = setup_volume(prereg, properties).await {
+                        warn!("Failed to setup volume: {e}");
+                    }
+                });
+            }
+            DeviceEvent::Remove(device) => registrations.remove(&device),
+        }
+    }
+}
+
+#[instrument(skip(r, properties))]
+async fn setup_volume(r: PreRegistration, properties: Properties) -> anyhow::Result<()> {
+    let devname = properties
+        .get("udev.DEVNAME")
+        .ok_or_else(|| anyhow::anyhow!("Failed to read udev property DEVNAME"))?;
+    let uuid = properties
+        .get("udev.ID_FS_UUID")
+        .ok_or_else(|| anyhow::anyhow!("Failed to read udev property ID_FS_UUID"))?;
+
+    let state_dir = fs::canonicalize(
+        systemd_directories::state_dir()
+            .or(dirs::state_dir())
+            .ok_or_else(|| anyhow::anyhow!("Failed to get systemd state directory"))?,
+    )
+    .await?;
+    let path = format!("{}/{}", state_dir.to_string_lossy(), uuid);
```

Comment:
don't make paths like this; use PathBuf::join

Reply:
Done.